Add an overview of the atomistic models facilities in the documentation (#541)
lab-cosmo/metatensor: docs/src/atomistic/overview.rst

Once a model is defined and trained, it should be exported by constructing a
:py:class:`MetatensorAtomisticModel`, and calling ``export`` on it. This class
is a wrapper for the model that will handle unit conversions on input and
outputs. It will also store metadata about the model (such as the authors, a
list of references, …) and the model capabilities (what properties it can
compute, which neighbors list the whole model requires, …). Optionally, this
class can also check that both data provided by the engine and properties
computed by the model follow the metatensor interface, which can be used to
debug your code.

Finally, the exported model can be loaded by simulation engines and used to run
simulations and make predictions.

.. _TorchScript: https://pytorch.org/docs/stable/jit.html

Constrains on atomistic models
------------------------------

There are a couple of constrains on what a given model must do to be useable
with metatensor, but appart from these you can do what you want inside the
model!

The main constrain is that the model must be compatible with `TorchScript`_,
i.e. you must use either pure PyTorch code in the definition of your model, or

> implement a custom TorchScript extension for any operations where a pure PyTorch

Reviewer: Is there a link on how to build these custom extensions?

implementation is too slow or too much work.

Your model should also only take data from our :py:class:`System` definition:
atomic types and positions, simulation cell, and list of neighbors for different
spherical cutoffs.

If you need additional data that you can not compute inside the model (such as
atomic spins, non-spherical particle orientation, *etc.*) you can use
:py:meth:`System.get_data`, but this comes with significant caveats:

> 

Reviewer: Maybe we should say before we introduce the `data` attribute that people should talk to us if they want a quantity in the `systems`.

Maybe with a link to open a guthub issue.

Author: This "contact us" is just below, do you mean it should be moved above? This would feel a bit weird, since readers don't yet know what we are taking about. I can add a link to github though

outputs (identified by the corresponding key in the output dictionary). These
standardized outputs are documented in :ref:`this page
<atomistic-models-outputs>`.

Data flow between the model and engine
--------------------------------------

The sequence of operations to use a metatensor atomistic model from a simulation
engine follows the same high level sequence of operations, illustrated and
explained below.

.. make the `tip` admonition grey only for this page
.. raw:: html

    <style>
        body[data-theme="light"] {
            --color-admonition-title--tip: #7c7c7c;
            --color-admonition-title-background--tip: #b9b9b9;
        }

        body[data-theme="auto"] {
            @media (prefers-color-scheme: light) {
                --color-admonition-title--tip: #7c7c7c;
                --color-admonition-title-background--tip: #b9b9b9;
            }
        }
    </style>

.. figure:: ../../static/images/model-dataflow.*
    :width: 600px
    :align: center

    Illustration of the flow of data between the engine and the model.

1. the engine loads an exported model from a file;

   .. tip::

        The engine should use :py:func:`check_atomistic_model` or

>         :cpp:func:`metatensor_torch::load_atomistic_model` to also perform

Reviewer: Is this also exposed to Python? We might also use this in metatensor-models. ☺️

Author: Yes, it is exposed to Python! https://lab-cosmo.github.io/metatensor/latest/atomistic/reference/models/index.html#metatensor.torch.atomistic.check_atomistic_model

        checks before loading the model.

> 2. the engine get the capabilities of the model from the model;

Reviewer: ```suggestion
2. the engine requests and gets the capabilities from the model;
```

   .. tip::

        This can be done by calling
        :py:func:`MetatensorAtomisticModel.capabilities`. This function is also
        exported to TorchScript and can be called from C++ with
        :cpp:func:`torch::jit::Module::run_method`.

3. the engine creates the :py:class:`ModelEvaluationOptions` based on the
   model's capabilities and user input;

> 4. the engine create :py:class:`System` matching it's own internal data

Reviewer: ```suggestion
4. the engine creates a :py:class:`System` matching it's own internal data
```

   representation;

   .. tip::
        The ``positions`` and ``cell`` should have their respective
        ``requires_grad`` parameters set if the engine wants to run backward
        propagation at step 10.

5. the engine asks the model for the required neighbors lists;

   .. tip::

        This can be done by calling
        :py:func:`MetatensorAtomisticModel.requested_neighbors_lists`. This
        function is also exported to TorchScript and can be called from C++ with
        :cpp:func:`torch::jit::Module::run_method`.

> 6. the engine compute the neighbors lists corresponding to the model requests,

Reviewer: ```suggestion
6. the engine computes the neighbors lists corresponding to the model requests,
```

   and register them with all systems;

   .. tip::

        If the engine does not use torch to compute the neighbors lists (using
        instead some other neighbors list implementation), the neighbors list
        should be registered with torch's automatic differentiation framework by
        using :py:func:`register_neighbors_autograd` before adding the neighbors
        lists to the systems.

> 7. the engine calls the model ``forward`` function with all the systems, the

Reviewer: ```suggestion
7. the engine calls the model ``forward()`` function with all the systems, the
```

    existing models: just load them in your simulation engine and hit the floor
    running!

Developing of simulation engines
    You are working on software for molecular simulations, including algorithms
    to sample different thermodynamic ensembles, or high performance simulation
    code.

    By using metatensor, you'll get access to the whole space of machine
    learning potentials at once! You'll also get to use models for more than
    predicting the energy of a system (for example using ML models for charge
    transfers, predicting polarizability along a trajectory, *etc.*).

How it works
------------

.. py:currentmodule:: metatensor.torch.atomistic

Metatensor atomistic models are based on PyTorch, and more particularly
`TorchScript`_. TorchScript is a programming language which is mainly a subset
of Python, and PyTorch contains a compiler from Python to TorchScript code.
After doing this translation, the model no longer depends on Python and can be
executed directly inside simulation engines implemented in C, C++, Fortran, …
This approach allow to define and tweak models as Python code, and then once

> they are working as intended, export them to a more stable representation to be

Author: I went with "Python-independant representation"

used in simulations.

In practice, models should be defined as custom :py:class:`torch.nn.Module`
sub-class, following our :py:class:`ModelInterface`. New models can be written
using this interface directly, and pre-existing models can use a small wrapper
to convert from this interface to the model's existing input and output data.
The models take as input a set of atomistic :py:class:`System` (typically a

> single one during simulations, and multiple systems during training); a set of

Author: I'm not sure what you mean here. One could need to do inference on multiple systems at the same time for PIMD, replica-exchange MD, … so the separation is not that clear cut.

in a dictionary of :py:class:`metatensor.torch.TensorMap`, one such tensor map
for each property (i.e. energy, atomic charges, dipole, electronic density,
chemical shielding, *etc.*)

Once a model is defined and trained, it should be exported by constructing a
:py:class:`MetatensorAtomisticModel`, and calling ``export`` on it. This class
is a wrapper for the model that will handle unit conversions on input and
outputs. It will also store metadata about the model (such as the authors, a
list of references, …) and the model capabilities (what properties it can
compute, which neighbors list the whole model requires, …). Optionally, this
class can also check that both data provided by the engine and properties
computed by the model follow the metatensor interface, which can be used to
debug your code.

Finally, the exported model can be loaded by simulation engines and used to run
simulations and make predictions.

.. _TorchScript: https://pytorch.org/docs/stable/jit.html

Constrains on atomistic models
------------------------------

There are a couple of constrains on what a given model must do to be useable

> with metatensor, but appart from these you can do what you want inside the

Reviewer 2: "apart"

        }
    </style>

.. figure:: ../../static/images/model-dataflow.*
    :width: 600px
    :align: center

    Illustration of the flow of data between the engine and the model.

1. the engine loads an exported model from a file;

   .. tip::

        The engine should use :py:func:`check_atomistic_model` or
        :cpp:func:`metatensor_torch::load_atomistic_model` to also perform
        checks before loading the model.

2. the engine requests and gets its capabilities from the model;

   .. tip::

        This can be done by calling
        :py:func:`MetatensorAtomisticModel.capabilities`. This function is also
        exported to TorchScript and can be called from C++ with
        :cpp:func:`torch::jit::Module::run_method`.

3. the engine creates the :py:class:`ModelEvaluationOptions` based on the
   model's capabilities and user input;

4. the engine creates a list of :py:class:`System` (typically the list only
   contains one system) matching it's own internal data representation;

   .. tip::
        The ``positions`` and ``cell`` should have their respective
        ``requires_grad`` parameters set if the engine wants to run backward
        propagation at step 10.

> 5. the engine asks the model for the required neighbors lists;

Reviewer 2: "neighbors list" see above
